File: iOS/Spaces V1.5/ViewModels/GlobalSearchViewModel.swift
import Combine
import FirebaseFirestore
import Foundation

@MainActor
final class GlobalSearchViewModel: ObservableObject {
    struct SpaceMessageSearchResult: Identifiable, Hashable {
        enum ConversationKind: Hashable {
            case space(Space)
            case ping(Ping)
        }

        let id: String
        let conversationKind: ConversationKind
        let title: String
        let subtitle: String
        let preview: String
        let createdAt: Date?
    }

    @Published var searchText = "" {
        didSet {
            refreshStaticResults()
            performMessageSearch()
        }
    }

    @Published private(set) var spaceResults: [Space] = []
    @Published private(set) var pingResults: [Ping] = []
    @Published private(set) var peopleResults: [PingParticipant] = []
    @Published private(set) var messageResults: [SpaceMessageSearchResult] = []
    @Published private(set) var isLoading = false
    @Published private(set) var currentUserID: String?

    private let spaceService: SpaceService
    private let pingService: PingService
    private var spaces: [Space]
    private var pings: [Ping] = []
    private var people: [PingParticipant] = []
    private var pingsListener: ListenerRegistration?
    private var searchTask: Task<Void, Never>?

    convenience init(
        spaces: [Space]
    ) {
        self.init(
            spaces: spaces,
            spaceService: SpaceService(),
            pingService: PingService()
        )
    }

    init(
        spaces: [Space],
        spaceService: SpaceService,
        pingService: PingService
    ) {
        self.spaces = spaces
        self.spaceService = spaceService
        self.pingService = pingService
        self.currentUserID = pingService.currentUserID()
        refreshStaticResults()
    }

    deinit {
        pingsListener?.remove()
        searchTask?.cancel()
    }

    func updateSpaces(_ spaces: [Space]) {
        self.spaces = spaces
        refreshStaticResults()
        performMessageSearch()
    }

    func start() {
        guard pingsListener == nil else { return }

        pingsListener = pingService.listenToPingsForCurrentUser { [weak self] result in
            guard let self else { return }
            Task { @MainActor in
                self.currentUserID = self.pingService.currentUserID()
                switch result {
                case .success(let pings):
                    self.pings = pings
                    self.refreshStaticResults()
                    self.performMessageSearch()
                case .failure:
                    self.pings = []
                    self.refreshStaticResults()
                }
            }
        }

        Task {
            let participants = (try? await pingService.fetchAvailableParticipants()) ?? []
            people = participants
            refreshStaticResults()
        }
    }

    private func refreshStaticResults() {
        let query = searchText.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !query.isEmpty else {
            spaceResults = []
            pingResults = []
            peopleResults = []
            messageResults = []
            isLoading = false
            searchTask?.cancel()
            searchTask = nil
            return
        }

        let normalized = query.folding(options: [.caseInsensitive, .diacriticInsensitive], locale: .current)
        spaceResults = spaces.filter {
            [$0.name, $0.description, $0.emoji].contains { value in
                value.folding(options: [.caseInsensitive, .diacriticInsensitive], locale: .current).contains(normalized)
            }
        }

        pingResults = pings.filter { ping in
            let title = ping.title(for: pingService.currentUserID())
            let emoji = ping.emoji(for: pingService.currentUserID())
            return [title, emoji].contains { value in
                value.folding(options: [.caseInsensitive, .diacriticInsensitive], locale: .current).contains(normalized)
            }
        }

        let pingParticipantIDs = Set(pings.compactMap { $0.otherParticipant(for: pingService.currentUserID())?.id })
        peopleResults = people.filter { participant in
            !pingParticipantIDs.contains(participant.id) &&
            [participant.displayName, participant.emojiAvatar].contains { value in
                value.folding(options: [.caseInsensitive, .diacriticInsensitive], locale: .current).contains(normalized)
            }
        }
    }

    private func performMessageSearch() {
        searchTask?.cancel()
        let query = searchText.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !query.isEmpty else {
            messageResults = []
            isLoading = false
            return
        }

        searchTask = Task { [weak self] in
            guard let self else { return }
            isLoading = true

            let normalizedQuery = query.folding(options: [.caseInsensitive, .diacriticInsensitive], locale: .current)
            var results: [SpaceMessageSearchResult] = []

            for space in spaces {
                if Task.isCancelled { return }
                let messages = (try? await spaceService.fetchRecentMessages(in: space, limit: 25)) ?? []
                for message in messages where !message.deleted {
                    let searchable = [
                        message.senderName,
                        message.text ?? "",
                        message.replyContext?.preview ?? "",
                        message.spaceLinks.map(\.searchableText).joined(separator: "\n")
                    ]
                        .joined(separator: "\n")
                        .folding(options: [.caseInsensitive, .diacriticInsensitive], locale: .current)
                    guard searchable.contains(normalizedQuery) else { continue }
                    results.append(
                        SpaceMessageSearchResult(
                            id: "space-\(space.id)-\(message.id)",
                            conversationKind: .space(space),
                            title: space.name,
                            subtitle: message.senderName,
                            preview: message.text?.trimmingCharacters(in: .whitespacesAndNewlines).nilIfEmpty
                                ?? message.spaceLinks.first?.title
                                ?? "Message",
                            createdAt: message.createdAt
                        )
                    )
                }
            }

            for ping in pings {
                if Task.isCancelled { return }
                let messages = (try? await pingService.fetchRecentMessages(in: ping, limit: 25)) ?? []
                for message in messages where !message.deleted {
                    let searchable = [
                        message.senderName,
                        message.text ?? "",
                        message.replyContext?.preview ?? "",
                        message.spaceLinks.map(\.searchableText).joined(separator: "\n")
                    ]
                        .joined(separator: "\n")
                        .folding(options: [.caseInsensitive, .diacriticInsensitive], locale: .current)
                    guard searchable.contains(normalizedQuery) else { continue }
                    results.append(
                        SpaceMessageSearchResult(
                            id: "ping-\(ping.id)-\(message.id)",
                            conversationKind: .ping(ping),
                            title: ping.title(for: pingService.currentUserID()),
                            subtitle: message.senderName,
                            preview: message.text?.trimmingCharacters(in: .whitespacesAndNewlines).nilIfEmpty
                                ?? message.spaceLinks.first?.title
                                ?? "Message",
                            createdAt: message.createdAt
                        )
                    )
                }
            }

            guard !Task.isCancelled else { return }

            messageResults = results.sorted { lhs, rhs in
                (lhs.createdAt ?? .distantPast) > (rhs.createdAt ?? .distantPast)
            }
            isLoading = false
        }
    }
}

private extension String {
    var nilIfEmpty: String? {
        isEmpty ? nil : self
    }
}

File: Android V1.6/app/src/main/java/com/arcinteractive/spaces/ui/screens/search/GlobalSearchViewModel.kt
package com.arcinteractive.spaces.ui.screens.search

import android.content.Context
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.arcinteractive.spaces.data.model.Ping
import com.arcinteractive.spaces.data.model.PingParticipant
import com.arcinteractive.spaces.data.model.Space
import com.arcinteractive.spaces.data.pings.PingService
import com.arcinteractive.spaces.data.spaces.SpaceService
import com.google.firebase.firestore.ListenerRegistration
import java.util.Date
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class GlobalSearchUiState(
    val query: String = "",
    val currentUserId: String? = null,
    val spaceResults: List<Space> = emptyList(),
    val pingResults: List<Ping> = emptyList(),
    val peopleResults: List<PingParticipant> = emptyList(),
    val messageResults: List<GlobalMessageSearchResult> = emptyList(),
    val isLoading: Boolean = false,
    val errorMessage: String? = null
)

data class GlobalMessageSearchResult(
    val id: String,
    val sourceType: GlobalMessageSourceType,
    val sourceId: String,
    val title: String,
    val subtitle: String,
    val preview: String,
    val createdAt: Date?
)

enum class GlobalMessageSourceType {
    Space,
    Ping
}

class GlobalSearchViewModel(
    private val spaceService: SpaceService = SpaceService(),
    private val pingService: PingService = PingService()
) : ViewModel() {
    private val _uiState = MutableStateFlow(GlobalSearchUiState())
    val uiState: StateFlow<GlobalSearchUiState> = _uiState.asStateFlow()

    private var spaces: List<Space> = emptyList()
    private var pings: List<Ping> = emptyList()
    private var people: List<PingParticipant> = emptyList()
    private var currentUserId: String? = null
    private var pingsListener: ListenerRegistration? = null
    private var searchJob: Job? = null

    fun start(context: Context, spaces: List<Space>) {
        this.spaces = spaces
        this.currentUserId = pingService.currentUserId(context)
        _uiState.update { it.copy(currentUserId = currentUserId) }
        if (pingsListener == null) {
            pingsListener = pingService.listenToPingsForCurrentUser(
                context = context,
                listenerKey = "search.pings"
            ) { result ->
                result.onSuccess { loadedPings ->
                    pings = loadedPings
                    refreshStaticResults()
                    searchMessages(context)
                }.onFailure { error ->
                    _uiState.update { it.copy(errorMessage = error.localizedMessage ?: "Unable to load Pings.") }
                }
            }
            viewModelScope.launch {
                people = runCatching { pingService.fetchAvailableParticipants(context) }.getOrDefault(emptyList())
                refreshStaticResults()
            }
        } else {
            this.spaces = spaces
            refreshStaticResults()
        }
    }

    fun updateQuery(context: Context, query: String) {
        _uiState.update { it.copy(query = query, errorMessage = null) }
        refreshStaticResults()
        searchMessages(context)
    }

    fun clearErrorMessage() {
        _uiState.update { it.copy(errorMessage = null) }
    }

    fun createOrOpenPing(context: Context, participant: PingParticipant, onComplete: (Ping?) -> Unit) {
        viewModelScope.launch {
            runCatching {
                pingService.createOrOpenPing(context, participant)
            }.onSuccess { ping ->
                onComplete(ping)
            }.onFailure { error ->
                _uiState.update { it.copy(errorMessage = error.localizedMessage ?: "Unable to open this Ping.") }
                onComplete(null)
            }
        }
    }

    private fun refreshStaticResults() {
        val query = _uiState.value.query.trim()
        if (query.isBlank()) {
            searchJob?.cancel()
            _uiState.update {
                it.copy(
                    spaceResults = emptyList(),
                    pingResults = emptyList(),
                    peopleResults = emptyList(),
                    messageResults = emptyList(),
                    isLoading = false
                )
            }
            return
        }

        val pingParticipantIds = pings.mapNotNull { it.otherParticipant(currentUserId)?.id }.toSet()
        _uiState.update {
            it.copy(
                spaceResults = spaces.filter { space ->
                    listOf(space.name, space.description, space.emoji).any { value ->
                        value.contains(query, ignoreCase = true)
                    }
                },
                pingResults = pings.filter { ping ->
                    listOf(ping.title(currentUserId), ping.emoji(currentUserId)).any { value ->
                        value.contains(query, ignoreCase = true)
                    }
                },
                peopleResults = people.filter { participant ->
                    participant.id !in pingParticipantIds &&
                        listOf(participant.displayName, participant.emojiAvatar).any { value ->
                            value.contains(query, ignoreCase = true)
                        }
                }
            )
        }
    }

    private fun searchMessages(context: Context) {
        searchJob?.cancel()
        val query = _uiState.value.query.trim()
        if (query.isBlank()) return

        searchJob = viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true) }
            val results = mutableListOf<GlobalMessageSearchResult>()

            spaces.forEach { space ->
                val messages = runCatching {
                    spaceService.fetchRecentMessages(context, space, limit = 25)
                }.getOrDefault(emptyList())
                messages.filter { !it.deleted }.forEach { message ->
                    val searchable = listOf(
                        message.senderName,
                        message.text.orEmpty(),
                        message.replyContext?.preview.orEmpty(),
                        message.spaceLinks.joinToString("\n") { it.title + "\n" + (it.subtitle ?: "") }
                    )
                        .joinToString("\n")
                    if (searchable.contains(query, ignoreCase = true)) {
                        results += GlobalMessageSearchResult(
                            id = "space-${space.id}-${message.id}",
                            sourceType = GlobalMessageSourceType.Space,
                            sourceId = space.id,
                            title = space.name,
                            subtitle = message.senderName,
                            preview = message.text?.trim().orEmpty().ifBlank { message.spaceLinks.firstOrNull()?.title ?: "Message" },
                            createdAt = message.createdAt
                        )
                    }
                }
            }

            pings.forEach { ping ->
                val messages = runCatching {
                    pingService.fetchRecentMessages(context, ping, limit = 25)
                }.getOrDefault(emptyList())
                messages.filter { !it.deleted }.forEach { message ->
                    val searchable = listOf(
                        message.senderName,
                        message.text.orEmpty(),
                        message.replyContext?.preview.orEmpty(),
                        message.spaceLinks.joinToString("\n") { it.title + "\n" + (it.subtitle ?: "") }
                    )
                        .joinToString("\n")
                    if (searchable.contains(query, ignoreCase = true)) {
                        results += GlobalMessageSearchResult(
                            id = "ping-${ping.id}-${message.id}",
                            sourceType = GlobalMessageSourceType.Ping,
                            sourceId = ping.id,
                            title = ping.title(currentUserId),
                            subtitle = message.senderName,
                            preview = message.text?.trim().orEmpty().ifBlank { message.spaceLinks.firstOrNull()?.title ?: "Message" },
                            createdAt = message.createdAt
                        )
                    }
                }
            }

            _uiState.update {
                it.copy(
                    messageResults = results.sortedByDescending { item -> item.createdAt?.time ?: Long.MIN_VALUE },
                    isLoading = false
                )
            }
        }
    }

    override fun onCleared() {
        pingsListener?.remove()
        pingsListener = null
        searchJob?.cancel()
        super.onCleared()
    }
}
